libraries/map/src/main/kotlin/io/element/android/libraries/map/MapView.kt

```diff
@@ -0,0 +1,263 @@
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.platform.LocalInspectionMode
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.viewinterop.AndroidView
+import com.mapbox.mapboxsdk.Mapbox
+import com.mapbox.mapboxsdk.camera.CameraPosition
+import com.mapbox.mapboxsdk.camera.CameraUpdateFactory
+import com.mapbox.mapboxsdk.geometry.LatLng
+import com.mapbox.mapboxsdk.maps.MapView
+import com.mapbox.mapboxsdk.maps.MapboxMap
+import com.mapbox.mapboxsdk.maps.Style
+import com.mapbox.mapboxsdk.plugins.annotation.SymbolManager
+import com.mapbox.mapboxsdk.plugins.annotation.SymbolOptions
+import io.element.android.libraries.designsystem.preview.ElementPreviewDark
+import io.element.android.libraries.designsystem.preview.ElementPreviewLight
+import io.element.android.libraries.designsystem.theme.ElementTheme
+import io.element.android.libraries.designsystem.theme.components.FloatingActionButton
+import io.element.android.libraries.designsystem.theme.components.Icon
+import io.element.android.libraries.map.internal.buildTileServerUrl
+import kotlinx.collections.immutable.ImmutableList
+import kotlinx.collections.immutable.toImmutableList
+import timber.log.Timber
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * Composable wrapper around MapLibre's [MapView].
+ */
+@SuppressLint("MissingPermission")
+@Composable
+fun MapView(
+    modifier: Modifier = Modifier,
+    mapState: MapState = rememberMapState(),
+    darkMode: Boolean = !ElementTheme.colors.isLight,
+    onLocationClick: () -> Unit,
+) {
+    // When in preview, early return a Box with the received modifier preserving layout
+    if (LocalInspectionMode.current) {
+        Box(modifier = modifier)
+        return
+    }
+
+    val context = LocalContext.current
+    val mapView = remember {
+        Mapbox.getInstance(context)
+        MapView(context)
+    }
+    var mapRefs by remember { mutableStateOf<MapRefs?>(null) }
+
+    // Build map
+    LaunchedEffect(darkMode) {
+        mapView.awaitMap().let { map ->
+            map.uiSettings.apply {
+                isCompassEnabled = false
+            }
+            map.setStyle(buildTileServerUrl(darkMode = darkMode)) { style ->
+                mapRefs = MapRefs(
+                    map = map,
+                    symbolManager = SymbolManager(mapView, map, style).apply {
+                        iconAllowOverlap = true
+                    },
+                    style = style
+                )
+            }
+        }
+    }
+
+    // Update state position when moving map
+    LaunchedEffect(mapRefs) {
+        mapRefs?.let {
+            // TODO: Are we actually leaking memory by not removing this listener?
```

Comment:
I don't think leaking is a problem (it'll go away when the map goes away), but are we going to end up adding multiple listeners if `mapRefs` changes? (Can it change once set? I guess if we toggle dark mode?)

Reply:
Indeed it's as you say. Gotta think of something...

Reply:
Pushed a change that removes listeners.